src/shaka_scheme/runtime/stdproc/pairs_and_lists.cpp

```diff
@@ -0,0 +1,428 @@
+                                    "an empty list");
+  }
+  DataPair dp{create_node(Data(
+      DataPair(args[0])))};
+  return {create_node(Data(dp.cdr()))};
+}
+
+//(set-car! ...)
+Args set_car(Args args) {
+  if (args[0]->get_type() != Data::Type::DATA_PAIR) {
+    throw InvalidInputException(10,
+                                "set-car!: Type of given argument is not "
+                                    "DATA_PAIR");
+  }
+  if (args.size() != 2) {
+    throw InvalidInputException(11,
+                                "set-car!: Invalid number of arguments given");
+  }
+  args[0]->get<DataPair>().set_car(args[1]);
+  return {create_unspecified()};
+}
+
+//(set-cdr! ...)
+Args set_cdr(Args args) {
+  if (args[0]->get_type() != Data::Type::DATA_PAIR) {
+    throw InvalidInputException(12,
+                                "set-cdr!: Type of given argument is not "
+                                    "DATA_PAIR");
+  }
+  if (args.size() != 2) {
+    throw InvalidInputException(13,
+                                "set-cdr!: Invalid number of arguments given");
+  }
+  args[0]->get<DataPair>().set_cdr(args[1]);
+  return {create_unspecified()};
+}
+
+//(null? ...)
+Args is_null(Args args) {
+  if (args.size() != 1) {
+    throw InvalidInputException(7,
+                                "null?: Invalid number of arguments for "
+                                    "procedure");
+  }
+  return {create_node(Data(Boolean(
+      args[0]->get_type() == Data::Type::NULL_LIST)
+  ))};
+}
+
+//(list? ...)
+Args is_list(Args args) {
+  if (args.size() != 1) {
+    throw InvalidInputException(8,
+                                "list?: Invalid number of arguments for "
+                                    "procedure");
+  }
+  return {create_node(Data(Boolean(
+      ((args[0]->get_type() == Data::Type::DATA_PAIR) &&
+          (args[0]->get<DataPair>().cdr()->get_type() == Data::Type::DATA_PAIR))
+  )))};
+}
+
+//(make-list ...)
+Args make_list(Args args) {
+  // If the args given are either empty or have too many arguments
+  //  Throw an error
+  if (args.size() > 2 || args.empty()) {
+    throw InvalidInputException(9,
+                                "make-list: Invalid number of arguments for"
+                                    " procedure");
+  }
+  // Create an args to hold arguments for list
+  Args list_args;
+  // If a single argument is given
+  if (args.size() == 1) {
+    // If the type of the given argument is not a number
+    //  Throw an error
+    if (args[0]->get_type() != Data::Type::NUMBER) {
+      throw InvalidInputException(15,
+                                  "make-list: Given argument is not of type "
+                                      "NUMBER");
+    }
+
+    // For the value of the given number
+    //  Add a null list at the end
+    for (int i = 0; i < args[0]->get<Number>(); i++) {
+      list_args.push_back(create_node(Data()));
+    }
+  }
+    // Else when 2 arguments are given
+  else {

```

Comment:
Make sure to make this invalid for cases of 3 arguments and more. I may be wrong, so reference R7RS here closely.

Reply:
On line 130 of pairs_and_lists.cpp, it throws an exception when there are more than 2 arguments given.